src/ToData.purs

```diff
@@ -1,37 +1,111 @@
 module ToData
-  ( class ToData
+  ( Day(..)
+  , class ConstrIndex
+  , class IsIndexed
+  , class ToData
+  , class ToDataArgs
+  , genericIndexFromConstructor
+  , genericToData
+  , indexFromConstructor
   , toData
+  , toDataArgs
+  , constrToIndex
   ) where
 
 import Prelude
 
+import Contract.Prelude (Tuple(..), genericShow)
 import Data.Array as Array
-import Data.BigInt (BigInt)
+import Data.BigInt (BigInt, fromInt)
 import Data.BigInt as BigInt
 import Data.Either (Either(Left, Right))
 import Data.Foldable (class Foldable)
+import Data.Generic.Rep as G
 import Data.List (List)
 import Data.Map (Map)
 import Data.Map as Map
 import Data.Maybe (Maybe(Just, Nothing))
 import Data.Profunctor.Strong ((***))
 import Data.Ratio (Ratio, denominator, numerator)
+import Data.Symbol (class IsSymbol, reflectSymbol, SProxy(..))
 import Data.Tuple.Nested (type (/\), (/\))
 import Data.UInt (UInt)
 import Helpers (uIntToBigInt)
 import Prim.TypeError (class Fail, Text)
+import Type.Proxy (Proxy(..))
 import Types.ByteArray (ByteArray)
 import Types.PlutusData (PlutusData(Constr, Integer, List, Map, Bytes))
 
+data Day = Mon | Tue | Wed | Thurs | Fri | Sat | Sun
+
+derive instance genericDay :: G.Generic Day _
+
+instance showDay :: Show Day where
+  show = genericShow
+
+instance dayToData :: ToData Day where
+  toData = genericToData
+
 class ToData (a :: Type) where
   toData :: a -> PlutusData
 
+class ConstrIndex (a :: Type) where
+  constrToIndex :: Proxy a -> Map String Int
+
+instance dayConstrIndex :: ConstrIndex Day where
+  constrToIndex _ = Map.fromFoldable [ Tuple "Mon" 0 ]
+
+-- Generic
+class ToDataArgs a where
+  toDataArgs :: a -> Array (PlutusData)
+
+instance toDataArgsNoArguments :: ToDataArgs G.NoArguments where
+  toDataArgs _ = []
+
+instance toDataArgsArgument :: ToData a => ToDataArgs (G.Argument a) where
+  toDataArgs (G.Argument x) = [ toData x ]
+
+instance toDataArgsProduct :: (ToDataArgs a, ToDataArgs b) => ToDataArgs (G.Product a b) where
+  toDataArgs (G.Product x y) = toDataArgs x <> toDataArgs y
+
+instance toDataSum' :: (ToDataArgs largs, ToDataArgs rargs) => ToData (G.Sum (G.Constructor lname largs) (G.Constructor rname rargs)) where
+  toData (G.Inl (G.Constructor args)) = Constr zero (toDataArgs args)
+  toData (G.Inr (G.Constructor args)) = Constr one (toDataArgs args)
+
+instance toDataSum'' :: (ToDataArgs args, ToData (G.Sum l r)) => ToData (G.Sum (G.Constructor name args) (G.Sum l r)) where
+  toData (G.Inl (G.Constructor args)) = Constr zero (toDataArgs args)
+  toData (G.Inr x) = toData x
```

Comment:
This is where I don't honestly know how to proceed. In order to set the `Constr` index field, I need to specify the indices exactly as the Haskell user did via `makeIsDataIndexed`. However, since we're operating on Generic.Rep, I don't have the entire information I need to perform such a resolution.

In essence, I only have the `name` of the constructor, however, I need the original type that this constructor belongs to. Any ideas?

Reply:
This apparently compiles, dunno if it works (too tired to figure out how to test it or throw together a PR), might be helpful.

```haskell
data Z
data S n

class Countable n where
  count :: Proxy n -> BigInt

instance countableZ :: Countable Z where
  count _ = fromInt 0

instance countableS :: (Countable n) => Countable (S n) where
  count _ = fromInt 1 + count (Proxy :: Proxy n)

class ToDataWithIndex t n where
  toDataWithIndex :: Countable n => Proxy n -> t -> PlutusData

instance toDataWithIndexSumA :: (ToDataArgs la, ToDataWithIndex (G.Sum a b) (S n), Countable n)
                                => ToDataWithIndex (G.Sum (G.Constructor ln la) (G.Sum a b)) n  where
  toDataWithIndex n (G.Inl (G.Constructor args)) = Constr (count n) (toDataArgs args)
  toDataWithIndex n (G.Inr x) = toDataWithIndex (Proxy :: Proxy (S n)) x

instance toDataWithIndexSumB :: (ToDataArgs la, ToDataArgs ra, Countable n)
                                => ToDataWithIndex (G.Sum (G.Constructor ln la) (G.Constructor rn ra)) n where
  toDataWithIndex n (G.Inl (G.Constructor args)) = Constr (count n) (toDataArgs args)
  toDataWithIndex n (G.Inr (G.Constructor args)) = Constr (count n + fromInt 1) (toDataArgs args)

instance (IsSymbol ln,ToDataWithIndex (G.Sum (G.Constructor ln la) (G.Sum a b)) Z)
         => ToData (G.Sum (G.Constructor ln la) (G.Sum a b)) where
  toData x = toDataWithIndex (Proxy :: Proxy Z) x

instance ToDataWithIndex (G.Sum (G.Constructor ln la) (G.Constructor rn ra)) Z
         => ToData (G.Sum (G.Constructor ln la) (G.Constructor rn ra)) where
  toData x = toDataWithIndex (Proxy :: Proxy Z) x
 ```

Reply:
Thanks @gnumonik, this is a good solution for when we want to have something like [unstableMakeIsData](https://github.com/input-output-hk/plutus/blob/1f31e640e8a258185db01fa899da63f9018c0e85/plutus-tx/src/PlutusTx/IsData/TH.hs#L129). However, given our Haskell compatibility requirement, this is a bit more trickier and we'd need an explicit control over which Constructor gets assigned which index.

In `unstableMakeIsData` the indices are increasing `[0..]` as one would expect, however, we're implicitly relying on the order of constructors as yielded by that [TH.datatypeCons](https://github.com/input-output-hk/plutus/blob/1f31e640e8a258185db01fa899da63f9018c0e85/plutus-tx/src/PlutusTx/IsData/TH.hs#L124).

Reply:
Hey @gnumonik could you help me with the following idea

```purescript
instance monConstrIndex :: ConstrIndex Day "Mon" where
   resolveIndex _ _ = zero

instance tueConstrIndex :: ConstrIndex Day "Tue" where
   resolveIndex _ _ = zero

class IsSymbol s <= ConstrIndex (a :: Type) (s :: Symbol) where
  resolveIndex :: Proxy a -> SProxy s -> BigInt


class ToDataWithIndex a ci where
  toDataWithIndex :: forall s. ConstrIndex ci s => Proxy ci -> a -> PlutusData

-- class ToDataWithIndex t where
--   toDataWithIndex :: BigInt -> t -> PlutusData

instance toDataWithIndexSumB :: (IsSymbol ln, IsSymbol rn, ToDataArgs la, ToDataArgs ra)
                                => ToDataWithIndex (G.Sum (G.Constructor ln la) (G.Constructor rn ra)) ci where
  toDataWithIndex n (G.Inl (G.Constructor args)) = Constr (resolveIndex n (SProxy :: SProxy ln)) (toDataArgs args)
  toDataWithIndex n (G.Inr (G.Constructor args)) = Constr (resolveIndex n (SProxy :: SProxy rn)) (toDataArgs args)

instance toDataWithIndexSumA :: (IsSymbol ln, ConstrIndex ci ln, ToDataArgs la, ToDataWithIndex (G.Sum a b) ci)
                                => ToDataWithIndex (G.Sum (G.Constructor ln la) (G.Sum a b)) ci where
  toDataWithIndex n (G.Inl (G.Constructor args)) = Constr (resolveIndex n (SProxy :: SProxy ln)) (toDataArgs args)
  toDataWithIndex n (G.Inr x) = toDataWithIndex (n :: Proxy ci) x


instance (ConstrIndex ci s, ToDataWithIndex (G.Sum l r) ci s) => ToData (G.Sum l r) where
   toData x = toDataWithIndex (Proxy :: Proxy ci) x
```

Reply:
I didn't manage to get it to compile, fails at

`toDataWithIndex n (G.Inr x) = toDataWithIndex (n :: Proxy ci) x`

Reply:
@gnumonik we made it work! Check out the latest commit